File: src/allowlist.ts
/**
 * allowlist.ts
 *
 * Permitted editor allowlist.
 * Sources (both are merged if provided):
 *   DOCUMENT_EDITORS      — comma-separated handles in env var
 *   DOCUMENT_EDITORS_FILE — path to a file with one handle per line
 *
 * OCI Functions note:
 *   Cache is per-invocation only (no shared memory between calls).
 *   For large lists, prefer DOCUMENT_EDITORS env var over file.
 */

import { normalizeHandle } from "./webfinger.ts";

let cachedAllowlist: Set<string> | null = null;

async function loadAllowlist(): Promise<Set<string>> {
  if (cachedAllowlist) return cachedAllowlist;

  const allowlist = new Set<string>();

  const envList = Deno.env.get("DOCUMENT_EDITORS");
  if (envList) {
    for (const handle of envList.split(",")) {
      const n = normalizeHandle(handle.trim());
      if (n) allowlist.add(n);
    }
  }

  const filePath = Deno.env.get("DOCUMENT_EDITORS_FILE");
  if (filePath) {
    try {
      const content = await Deno.readTextFile(filePath);
      for (const line of content.split("\n")) {
        const n = normalizeHandle(line.trim());
        if (n) allowlist.add(n);
      }
    } catch (err) {
      console.error(`Failed to load allowlist from ${filePath}:`, err);
    }
  }

  cachedAllowlist = allowlist;
  return allowlist;
}

export async function validateAllowlist(handle: string): Promise<boolean> {
  const allowlist = await loadAllowlist();
  return allowlist.has(normalizeHandle(handle));
}

export function resetAllowlistCache(): void {
  cachedAllowlist = null;
}

export async function getAllowlist(): Promise<string[]> {
  return Array.from(await loadAllowlist()).sort();
}

File: lib/koto_cms/allowlist.ex
defmodule KotoCms.Allowlist do
  def validate(handle) do
    allowlist = load_allowlist()
    normalized = normalize_handle(handle)
    MapSet.member?(allowlist, normalized)
  end

  def get_all do
    load_allowlist() |> MapSet.to_list() |> Enum.sort()
  end

  defp load_allowlist do
    env_list = System.get_env("DOCUMENT_EDITORS") || ""
    file_path = System.get_env("DOCUMENT_EDITORS_FILE")

    env_handles =
      env_list
      |> String.split(",")
      |> Enum.map(&String.trim/1)
      |> Enum.reject(&(&1 == ""))
      |> Enum.map(&normalize_handle/1)

    file_handles =
      if file_path do
        case File.read(file_path) do
          {:ok, content} ->
            content
            |> String.split("\n")
            |> Enum.map(&String.trim/1)
            |> Enum.reject(&(&1 == ""))
            |> Enum.map(&normalize_handle/1)

          {:error, _} -> []
        end
      else
        []
      end

    MapSet.new(env_handles ++ file_handles)
  end

  defp normalize_handle(handle) do
    handle
    |> String.replace_prefix("@", "")
    |> String.downcase()
  end
end
